### PR TITLE
Add ability to start websocket connection with a frame size or a Codec

### LINKED ISSUE #1873: expose some way to alter websocket frame size limit

### PR DESCRIPTION
**Pull request by sazid**
## PR Type

Feature

## PR Checklist

- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview

**New**: You can now start a websocket session with a frame size or a `actix_http::ws::Codec` by calling either `ws::start_with_frame_size()` or `ws::start_with_codec()`.

Fixes #1873

### CODE REVIEW

File: actix-web-actors/src/ws.rs [hunk 1]
```diff
@@ -40,6 +40,43 @@ where
...
+    let mut res = handshake(req)?;
+    Ok(res.streaming(WebsocketContext::with_codec(
+        actor,
+        stream,
+        Codec::new().max_size(frame_size),
+    )))
+}
+
+/// Perform WebSocket handshake and start actor with the given
+/// `actix_http::ws::Codec`.
```
Comment (robjtede) [inline actix-web-actors/src/ws.rs]: use intra-doc links

```suggestion
/// Perform WebSocket handshake and start actor with the given [`Codec`].
```

**Review by therealprof — COMMENTED** [review]
Nice. But since we're touching this it would be great to also add the possibility to interact with the handshake. Up to now I had to do the WS handshake myself a couple of times, one reason being the insufficient `max_size` (as addressed by this PR) and the other being the abilitity to introspect and set headers as well as dealing with WebSocket extensions.

**Comment by robjtede** [comment]
We've had a chat on Gitter and are looking at a builder style solution to cover more of the custom cases. Either way this feature will be in v4.

**Comment by therealprof** [comment]
> We've had a chat on Gitter and are looking at a builder style solution to cover more of the custom cases. Either way this feature will be in v4.

Excellent, maybe we should close this PR then to avoid confusion?

**Comment by robjtede** [comment]
> Excellent, maybe we should close this PR then to avoid confusion?

I think @sazid is gonna tackle it in this PR. Correct me if I'm wrong. But I will mark as draft for now.

**Comment by therealprof** [comment]
Marking as draft would have been my next suggestion, thanks. 😅

**Comment by sazid** [comment]
Oops, I think I didn't read the last two comments.

**Comment by sazid** [comment]
Reopened a new PR with builder approach. 
https://github.com/actix/actix-web/pull/1920

**Pull request closed without merging**